### PR TITLE
feat: add `validate` flag to statically check configuration files

### LINKED ISSUE #293: Feat:  add `beelzebub validate` CLI command for static configuration checks
## Summary

When I configure Beelzebub I'd like to have a static way to predict the outcome of my configurations, without it I ran against the following failures:

**Missing `protocol`**: the daemon crashes with `"protocol  not managed"`. You get a fatal error, but it does not tell you which file is wrong.

**Empty `address`**: the daemon **starts without any error**. `http.ListenAndServe("")` binds to port 80 by default. You think you configured `:9090`, but you are listening on `:80` and there is no warning.

**`LLMHoneypot` without `llmProvider`**: the daemon **starts without any error**. I tested this end-to-end with a real LLM backend (NVIDIA API). With a valid `llmProvider: "openai"`, the honeypot returns realistic, varied responses like:
```
GET /wp-admin  returns  "405 Method Not Allowed — You are accessing a restricted area."
GET /secret    returns  "403 Forbidden — You do not have permission. Apache/2.4.41"
```
Without `llmProvider`, the same honeypot returns `"500 Internal Server Error"` to every request, the risk if to not gain the attention of the attacker, turning the trap into an open door.

**Two services on the same port**: the daemon **starts without any error**. The first service wins, the second logs a single `"error during init HTTP Protocol: bind: address already in use"` at error level and continues.

## Proposed Solution

A `--validate` flag that checks all configs and exits. No network listeners. No daemon. Similar to other existing and world-wide adopted tools, like the flag `-t` for `nginx` for example.

```bash
beelzebub --validate --confServices ./configurations/services/

✓ ssh-2222.yaml    — SSH honeypot on :2222, 1 command
✓ http-8080.yaml   — HTTP honeypot on :8080, 1 command
✗ http-8081.yaml   — ERROR: missing required field "protocol"
✗ ssh-llm.yaml     — ERROR: plugin LLMHoneypot requires llmProvider and llmModel
⚠ ssh-2222.yaml    — WARNING: openAISecretKey is set inline, consider using env var

2 errors, 1 warning
```

Checks:

| Check | Severity |
|---|---|
| `protocol` present and valid (http, ssh, tcp, mcp, telnet) | Error |
| `address` not empty | Error |
| `LLMHoneypot` plugin without `llmProvider` or `llmModel` | Error |
| Broken regex in `commands` | Error |
| Two services on the same protocol + address | Error |
| `deadlineTimeoutSeconds` not set | Warning |

The change should be small and self-contained from my first analysis, but beware I'm also quite dizzy from the day of work, therefore please show me if I made any mistake

- `main.go` add the flag and early exit
- `parser/validator.go` new file with validation logic
- `parser/validator_test.go` new file, table-driven tests

I expect existing parser types and DI pattern to be reused as-is, no changes to builder, protocols, plugins, or tracing are expected. 

## Alternatives Considered

JSON Schema validation? adds a dependency and a schema to keep in sync with Go structs but the structs are already the source of truth, no need for a second one and doesn't feel like the `Go` way.

Strict YAML unmarshalling, would catch unknown fields but cannot express cross-field rules like "LLMHoneypot needs llmProvider" or cross-file rules like port collisions.

Pre-commit hooks, is a nice to have but doesn't help in CI/CD or Docker deployments where configs are mounted at runtime.

## Use Cases

As a security engineer, I want to validate configs before deploying so I do not accidentally end up running a honeypot that looks alive but is not actually catching anything.

As a DevOps engineer, I want a dry-run command I can put in a CI pipeline to verify config changes before they hit production.

## Additional Context

- No static validation exists in the codebase today,apart from what happens at runtime inside `ReadConfigurationsServices()` (regex compilation, rate limit validation) or inside `builder.Run()` (protocol switch).
- Valid protocols are defined in `builder/builder.go:153-165`: http, ssh, tcp, mcp, telnet.
- This is the same pattern as `nginx -t`, `haproxy -c`, `helm lint`, `kubectl apply --dry-run`.

## Acceptance Criteria

- [x] `--validate` parses all YAML in the services directory without starting any listener
- [x] Catches: missing/invalid protocol, empty address, LLMHoneypot without llmProvider/llmModel, broken regex, port collisions
- [x] Warns about: inline openAISecretKey, missing deadlineTimeoutSeconds
- [x] Exit code 0 on success, 1 on any error
- [x] Tests cover all rules using the existing mock/DI pattern
- [x] All existing tests still pass

@mariocandela if it's okay and is within the direction of the project I could work on this during the weekend and the beginning of the upcoming week, however I see that there are some open pull requests, should I wait for the merge of said pull requests before starting?

### LINKED ISSUE #296: bug: flaky unit test due to data race
The unit tests occasionally fail on `TestVerifyConfigurationsChanged` when running tests with the `-race` flag enabled, like how we do it in CI.

The failure is caused by a data race between:

* an ongoing HTTP request (via `resty` / `net/http`)
* and `httpmock.DeactivateAndReset()` being called while the request is still in progress

This happens because the test starts a background goroutine (`verifyConfigurationsChanged`) that continues running after the test completes, while `httpmock` modifies global HTTP state during teardown.

#### Symptoms

* Intermittent CI failures
* `WARNING: DATA RACE` in logs
* Test passes locally sometimes, fails in CI

#### Root cause

* Background goroutine is not stopped before test teardown
* `httpmock` uses global state → unsafe with concurrent requests
* Shared variable (`callCount`) is not synchronized

#### Suggested fixes

* Introduce a stop mechanism for `verifyConfigurationsChanged` (e.g. channel or context)
* Ensure goroutines are terminated before `httpmock.DeactivateAndReset()`
* Make shared counters thread-safe (e.g. `sync/atomic`)
* (Optional) Refactor loop logic into a testable synchronous function

This is not just a simple test flakiness but a real concurrency issue surfaced by the race detector.

### PR DESCRIPTION
**Pull request by gzigurella**
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally before submission?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

## Summary

Adds a configuration validation engine (`beelzebub validate`) that reports all errors and warnings across core and service configs before startup, replacing the basic listing that existed in the plugin-system branch.

**Closes #293** — Feat: add `beelzebub validate` CLI command for static configuration checks

**Closes #296** — Bug: flaky unit test due to data race (context-based stop mechanism)

> **Note:** This PR supersedes #295 which was closed after the `plugin-system` branch was merged into `main` via #294. This branch is rebased onto `main` and includes additional fixes and improvements from my code review after rebase.

### Core engine (`internal/parser/validator.go`)
- `Validate()` for services (protocol, address, port, commands, plugins, TLS, collisions)
- `ValidateCore()` for core config (RabbitMQ, Cloud, Prometheus, logs)
- `ServiceValidator` interface with `init()`-based registration
- Shared helpers (`internal/parser/validator_shared.go`): `ValidateTLSConfig()` and `ValidatePasswordRegex()` used by protocol validators to avoid duplication

### Lenient parsing
- `ReadConfigurationsServicesForValidation()` collects all parse errors as `ValidationIssue` instead of failing on first
- `Filename` field added to `BeelzebubServiceConfiguration` for validation reporting

### Protocol validators
- **SSH**: passwordRegex (required, valid regex)
- **TELNET**: passwordRegex (required, valid regex)
- **HTTP**: TLS config (both cert+key or neither, file existence), fallback command
- **MCP**: tools validation (name, parameters)
- **TCP**: TLS config (shared with HTTP via `ValidateTLSConfig`)

### Plugin validators
- **LLMHoneypot**: provider (ollama/openai), model, secretKey
- **MazeHoneypot**: protocol (http)

### CLI (`cli/validate.go`)
- Rewritten to use the full validation engine
- Core config parse errors (e.g. malformed YAML) are now properly reported instead of being silently swallowed

## Fixes applied after rebase

- **Bug fix**: core config parse errors were silently discarded, causing `beelzebub validate` to report "0 errors" for unreadable configs
- **Refactor**: deduplicated TLS validation logic between HTTP and TCP into shared `ValidateTLSConfig()` helper
- **Refactor**: deduplicated passwordRegex validation logic between SSH and TELNET into shared `ValidatePasswordRegex()` helper
- **Fix**: removed duplicate `openAISecretKey` warning — the LLM plugin validator already handles provider-specific key validation
- **Style**: added doc comments on all exported types/functions, consistent with `configurations_parser.go`
- **Cosmetic**: fixed collision message format (`ssh::22` → `ssh :22`)
- **Data race fix**: replaced `done` channel with `context.Context` stop mechanism in `beelzebubCloud` — `Stop()` cancels context, goroutine exits cleanly before `httpmock.DeactivateAndReset()`, and `callCount` uses `atomic.Int64`

## Tests
~800+ lines of new tests across 20 test files, `go vet` clean, `go build` clean.
All existing tests continue to pass locally.

| Package | Tests |
|---------|-------|
| `cli/validate_test.go` | 6 tests for the CLI entry point (valid configs, invalid protocol, malformed core, missing core, malformed service, empty dir) |
| `internal/parser/validator_test.go` | 15+ test cases for core validation engine |
| `internal/parser/validator_shared_test.go` | 9 tests for shared helpers |
| `internal/protocols/*/validator_test.go` | Protocol-specific validator tests |
| `internal/plugins/*_validator_test.go` | Plugin-specific validator tests |

End-to-end tested with `beelzebub validate` against real and edge-case configurations.

### CODE REVIEW

**Comment by gzigurella** [comment]
Coverage is now properly addressed 🚀  As anticipated in #296, I also fixed the data race bug since I hit it while working on this PR, the changes include extracting `checkConfigurationsChanged()` as a testable synchronous function and adding a proper stop mechanism for the polling goroutine.

@airscripts @mariocandela I'll leave this for your review. There is of course no rush and I would also be happy to discuss it on our planned call in a couple of days if you prefer since the PR has grown a bit... (Oops 😛 )

File: internal/parser/validator.go [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+		if !ok {
+			r = &ValidationResult{Filename: issue.Filename}
+			resultMap[issue.Filename] = r
+		}
+		r.Issues = append(r.Issues, issue)
+	}
+
+	for i := range services {
+		r := getResult(resultMap, services[i].Filename)
+		services[i].Address = strings.TrimSpace(services[i].Address)
```
Comment (airscripts) [inline internal/parser/validator.go:95]: We normalize services[i].Address inside Validate(), but detectCollisions() still receives the original services slice, so collision detection is using the unnormalized addresses.  
That means two configs like ":8080" and " :8080 " will validate the address format similarly, but may not be flagged as duplicates.  

I think we should either normalize into the original slice before collision detection, or make detectCollisions() normalize internally so validation and collision checks operate on the same canonical value.

File: cli/validate.go [hunk 2]
```diff
@@ -2,74 +2,75 @@ package cli
...
-	"http": true, "ssh": true, "tcp": true, "telnet": true, "mcp": true,
+func validateConfigurations(_ *cobra.Command, _ []string) error {
+	return runValidate(rootConfCore, rootConfServices)
 }
 
-func validateConfigurations(_ *cobra.Command, _ []string) error {
-	// Let log level be controlled by rootLogLevel, don't force ErrorLevel here
-	// unless user didn't specify, but root command handles setting default.
+func runValidate(corePath, servicesPath string) error {
+	log.SetLevel(log.ErrorLevel)
```
Comment (airscripts) [inline cli/validate.go]: This unconditionally overrides the root --log-level flag for the validate command. The previous implementation explicitly avoided doing that, and root.go already handles log level setup centrally. Unless the goal is to intentionally ignore user-specified verbosity for validate, I’d avoid resetting the logger here and let the root command control it consistently.

File: internal/parser/configurations_parser.go [hunk 3]
```diff
@@ -237,87 +238,142 @@ func isNotFound(err error) bool {
...
-		return nil, fmt.Errorf("invalid BEELZEBUB_SERVICES_CONFIG: %v", err)
+		return nil, nil, fmt.Errorf("invalid BEELZEBUB_SERVICES_CONFIG: %v", err)
 	}
 
+	var issues []ValidationIssue
+	var validServices []BeelzebubServiceConfiguration
+
 	for i := range servicesConfiguration {
 		svc := &servicesConfiguration[i]
+		svc.Filename = "<env:BEELZEBUB_SERVICES_CONFIG>"
```
Comment (airscripts) [inline internal/parser/configurations_parser.go]: For env-backed configs, every service gets the same synthetic filename. That means multiple invalid entries from BEELZEBUB_SERVICES_CONFIG collapse into a single report bucket, and collisions/issues become hard to attribute to a specific array element. I’d consider using something indexed, like <env:BEELZEBUB_SERVICES_CONFIG>[0], so the output preserves per-entry attribution.

**Review by airscripts — COMMENTED** [review]
Other than those important threads I've opened, in terms of code quality and maintainability having also block separation with newlines where possible to make the code breathe and use constants (locally or globally) so code robustness is increased is a good practice to have in the long term.

File: internal/parser/validator.go (hunk 1 above)
Comment (gzigurella) [reply internal/parser/validator.go:95]: Yeah, I agree 😄 

The loop at line 93 does normalize addresses via strings.TrimSpace before
detectCollisions runs at line 106, so collision detection _currently_ works correctly.

But still you are right that detectCollisions should not rely on the caller to normalize first, I will add TrimSpace inside detectCollisions to make it self-contained and overall more robust.

File: cli/validate.go (hunk 2 above)
Comment (gzigurella) [reply cli/validate.go]: Good catch. I will remove it and let PersistentPreRunE handle the log level consistently.

The validation output is structured via Print() (fmt.Printf), and any logrus warnings are already suppressed in lenient mode.

File: internal/parser/configurations_parser.go (hunk 3 above)
Comment (gzigurella) [reply internal/parser/configurations_parser.go]: Yeah, I will use indexed filenames like <env:BEELZEBUB_SERVICES_CONFIG>[0], [1], etc. for per-entry attribution.

File: cli/validate.go (hunk 2 above)
Comment (gzigurella) [reply cli/validate.go]: I'll add a little more context, originally I thought that since the main focus defined inside #293 was to use the validation for a CI enforcing error-level was a good choice at the time, however now I'm thinking that there are occasions where user may want to run validate with debug level.

**Comment by gzigurella** [comment]
@airscripts all points should have been addressed as suggested 👍🏻 main branch has been merged cleanly as well after my changes 🚀

**Review by mariocandela — APPROVED** [review]
LGTM

**Comment by mariocandela** [comment]
Nice work @gzigurella ❤️
